sarc/cache.py

```diff
@@ -64,6 +66,199 @@ class CacheException(Exception):
     pass
 
 
+def ensure_utc(d: datetime) -> datetime:
+    assert d.tzinfo is not None
+    return d.astimezone(UTC)
+
+
+class CacheEntry:
+    """Describe a single cache entry at a point in time.
+
+    The cache entry contains multiple key-value pairs."""
+
+    _zf: ZipFile
+
+    def __init__(self, zf: ZipFile):
+        self._zf = zf
+
+    def add_value(self, key: str, value: bytes) -> None:
+        """Add a key-value pair to the cache entry"""
+        self._zf.writestr(key, value)
+
+    def get_value(self, key: str) -> bytes:
+        """Get the value for an existing key in this cache entry"""
+        return self._zf.read(key)
+
+    def get_keys(self) -> list[str]:
+        """Get the list of keys in the order they were added."""
+        return self._zf.namelist()
+
+    def close(self) -> None:
+        """Close the cache entry. MUST be called for new entries."""
```

Comment:
Since we must absolutely call `close()`, maybe this class should also be used as a context (with-statement) ?